Reduce the amount of String allocations in `updateLayersForInteractionRegions`

etiennesegonzac opened the pull request:
#### 5521eb299be3eb908c1855efbf1b2234900c4410
<pre>
Reduce the amount of String allocations in `updateLayersForInteractionRegions`
<a href="https://bugs.webkit.org/show_bug.cgi?id=292286">https://bugs.webkit.org/show_bug.cgi?id=292286</a>
&lt;<a href="https://rdar.apple.com/150275553">rdar://150275553</a>&gt;

Reviewed by Abrar Rahman Protyasha.

Instead of storing the InteractionRegion group name (String) on the
layer for reuse, use the ElementIdentifier&apos;s raw value instead.
The group name has been only derived from the ElementIdentifier for a
while.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::createInteractionRegionLayer):
(WebKit::interactionRegionElementIdentifierForLayer):
Store / read the identifier instead of the group name on the layer.
(WebKit::interactionRegionGroupNameForLayer): Deleted.
(WebKit::updateLayersForInteractionRegions):
Build the reusable layers hash map without Strings.
Fix a small double hashing issue.
Update the reuse code.

Canonical link: <a href="https://commits.webkit.org/294526@main">https://commits.webkit.org/294526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fa7d5d0d541da004bb4873462c19cd90b2361cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101368 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106521 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51997 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29526 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77211 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34246 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104375 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16467 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91559 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57552 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16287 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51345 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86167 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9645 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108873 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28497 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20969 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86185 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85743 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21955 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30468 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8178 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22614 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33707 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28239 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31559 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29798 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
```diff
@@ -152,7 +152,7 @@ static void applyBackgroundColorForDebuggingToLayer(CALayer *layer, const WebCor
 }
 
-static CALayer *createInteractionRegionLayer(WebCore::InteractionRegion::Type type, NSString *groupName)
+static CALayer *createInteractionRegionLayer(WebCore::InteractionRegion::Type type, WebCore::ElementIdentifier identifier, NSString *groupName)
```
It's not great to use `ElementIdentifier` in the UI process, since, in a future world with site isolation, the same layer tree will contain layers from separate web processes with conflicting elementIDs.

etiennesegonzac replied:
Good to know!
The code here is scoped a single RLT node without descending, so we should be safe. But other things might break... Filing it!

etiennesegonzac committed:
Reduce the amount of String allocations in `updateLayersForInteractionRegions`
https://bugs.webkit.org/show_bug.cgi?id=292286
<rdar://150275553>

Reviewed by Abrar Rahman Protyasha.

Instead of storing the InteractionRegion group name (String) on the
layer for reuse, use the ElementIdentifier's raw value instead.
The group name has been only derived from the ElementIdentifier for a
while.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::createInteractionRegionLayer):
(WebKit::interactionRegionElementIdentifierForLayer):
Store / read the identifier instead of the group name on the layer.
(WebKit::interactionRegionGroupNameForLayer): Deleted.
(WebKit::updateLayersForInteractionRegions):
Build the reusable layers hash map without Strings.
Fix a small double hashing issue.
Update the reuse code.

Canonical link: https://commits.webkit.org/294526@main
